debug.md

```diff
@@ -0,0 +1,150 @@
+# Debugging tips and tricks
+
+While working with the Go Client libraries you may run into some situations
+where you need a deeper level of understanding about what is going on in order
+to solve your problem. Here are some tips and tricks that you can use in these
+cases.
+
+## HTTP based clients
+
+All of our auto-generated clients have a constructor to create a client that
+uses HTTP/JSON instead of gRPC. Additionally a couple of our hand-written
+clients like Storage and Bigquery are also HTTP based. Here are some tips for
+debugging these clients.
+
+### Try setting Go's HTTP debug variable
+
+Try setting the following environment variable for verbose Go HTTP logging:
+GODEBUG=http2debug=1. To read more about this feature please see the godoc for
+[net/http](https://pkg.go.dev/net/http).
+
+### Add in your own logging with an HTTP middleware
+
+You may want to add in your own logging around HTTP requests. One way to do this
+is to register a custom HTTP client with a logging transport built in. Here is
+an example of how you would do this with the storage client.
+
+```go
+package main
+
+import (
+    "context"
+    "fmt"
+    "log"
+    "net/http"
+    "net/http/httputil"
+
+    "cloud.google.com/go/storage"
+    "google.golang.org/api/iterator"
+    "google.golang.org/api/option"
+    htransport "google.golang.org/api/transport/http"
+)
+
+type loggingRoundTripper struct {
+    rt http.RoundTripper
+}
+
+func (d loggingRoundTripper) RoundTrip(r *http.Request) (*http.Response, error) {
+    // Will create a dump of the request and body.
+    dump, err := httputil.DumpRequest(r, true)
+    if err != nil {
+        log.Println("error dumping request")
+    }
+    log.Printf("%s", dump)
+    return d.rt.RoundTrip(r)
+}
+
+func main() {
+    ctx := context.Background()
+
+    // Create a transport with authentication built-in detected with
+    // [ADC](https://google.aip.dev/auth/4110). Note you will have to pass any
+    // required scoped for the client you are using.
+    trans, err := htransport.NewTransport(ctx,
+        http.DefaultTransport,
```

Comment:
Note that this leaves out the perf tuning on the transport that we do in google-api-go-client. I've struggled to communicate this with customers in the past, any thoughts?

Reply:
I think that is fine for the scope of this document. I would say these are one-off type of things users should add to their code when required. I would not expect this to be something that is long-lived. So I think it is fine that we gloss over those details here. Thoughts?

Reply:
If that's the case I would want to clarify that these are intended for local debugging and not for production use. I have seen users put this kind of logic into production to try to catch transient issues for example. It's problematic if they accidentally hurt their performance while doing that

Reply:
Added a general top-level warning about perf and such